core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,589 @@
+    decodeList    = do txt <- D.decodeString
+                       return (toString txt) -- unpack lazily
+
+----------------------------------------------------------------------------
+-- Numeric data
+----------------------------------------------------------------------------
+
+instance Bi Integer where
+    encode = E.encodeInteger
+    decode = D.decodeInteger
+
+instance Bi Word where
+    encode = E.encodeWord
+    decode = D.decodeWord
+
+instance Bi Word8 where
+    encode = E.encodeWord8
+    decode = D.decodeWord8
+
+instance Bi Word16 where
+    encode = E.encodeWord16
+    decode = D.decodeWord16
+
+instance Bi Word32 where
+    encode = E.encodeWord32
+    decode = D.decodeWord32
+
+instance Bi Word64 where
+    encode = E.encodeWord64
+    decode = D.decodeWord64
+
+instance Bi Int where
+    encode = E.encodeInt
+    decode = D.decodeInt
+
+instance Bi Float where
+    encode = E.encodeFloat
+    decode = D.decodeFloat
+
+instance Bi Int32 where
+    encode = E.encodeInt32
+    decode = D.decodeInt32
+
+instance Bi Int64 where
+    encode = E.encodeInt64
+    decode = D.decodeInt64
+
+instance Bi Nano where
+    encode (MkFixed resolution) = E.encodeInteger resolution
+    decode = MkFixed <$> D.decodeInteger
+
+instance Bi Void where
+    decode = fail "instance Bi Void: you shouldn't try to deserialize Void"
+    encode = error "instance Bi Void: you shouldn't try to serialize Void"
+
+----------------------------------------------------------------------------
+-- Tagged
+----------------------------------------------------------------------------
+
+instance Bi a => Bi (Tagged s a) where
+    encode (Tagged a) = encode a
+    decode = Tagged <$> decode
+
+----------------------------------------------------------------------------
+-- Containers
+----------------------------------------------------------------------------
+
+instance (Bi a, Bi b) => Bi (a,b) where
+    encode (a,b) = E.encodeListLen 2
+                <> encode a
+                <> encode b
+    decode = do D.decodeListLenOf 2
+                !x <- decode
+                !y <- decode
+                return (x, y)
+
+instance (Bi a, Bi b, Bi c) => Bi (a,b,c) where
+    encode (a,b,c) = E.encodeListLen 3
+                  <> encode a
+                  <> encode b
+                  <> encode c
+
+    decode = do D.decodeListLenOf 3
+                !x <- decode
+                !y <- decode
+                !z <- decode
+                return (x, y, z)
+
+instance (Bi a, Bi b, Bi c, Bi d) => Bi (a,b,c,d) where
+    encode (a,b,c,d) = E.encodeListLen 4
+                    <> encode a
+                    <> encode b
+                    <> encode c
+                    <> encode d
+
+    decode = do D.decodeListLenOf 4
+                !a <- decode
+                !b <- decode
+                !c <- decode
+                !d <- decode
+                return (a, b, c, d)
+
+instance Bi BS.ByteString where
+    encode = E.encodeBytes
```

Comment:
Unfortunately I cannot comment on most of this code as it was written by Andrzej which is now on annual leave, so I'll try to give my best guess but I guess we'll need to take a leap of faith and clarify any obscure points when he gets back 😉